sale_invoice_policy/models/sale_order_line.py

```diff
@@ -5,64 +5,25 @@
 
 
 class SaleOrderLine(models.Model):
-
     _inherit = "sale.order.line"
 
     @api.depends(
         "qty_invoiced",
         "qty_delivered",
         "product_uom_qty",
-        "order_id.state",
-        "order_id.invoice_policy",
-    )
-    def _compute_qty_to_invoice(self):
-        invoice_policies = set(self.mapped("order_id.invoice_policy"))
-        line_by_id = {line.id: line for line in self}
-        done_lines = self.env["sale.order.line"].browse()
-        for invoice_policy in invoice_policies:
-            so_lines = (
-                self.with_context(invoice_policy=invoice_policy)
-                .filtered(lambda x, p=invoice_policy: x.order_id.invoice_policy == p)
-                .with_prefetch()
-            )
-            if so_lines:
-                done_lines |= so_lines
-                super(SaleOrderLine, so_lines)._compute_qty_to_invoice()
-                for line in so_lines:
-                    # due to the change of context in compute methods,
-                    # assign the value in the modified context to self
-                    line_by_id[line.id].qty_to_invoice = line.qty_to_invoice
-        # Not to break function if (it could not happen) some records
-        # were not in so_lines
-        super(SaleOrderLine, self - done_lines)._compute_qty_to_invoice()
-        return True
-
-    @api.depends(
         "state",
-        "product_uom_qty",
-        "qty_delivered",
-        "qty_to_invoice",
-        "qty_invoiced",
         "order_id.invoice_policy",
     )
-    def _compute_invoice_status(self):
-        invoice_policies = set(self.mapped("order_id.invoice_policy"))
-        line_by_id = {line.id: line for line in self}
-        done_lines = self.env["sale.order.line"].browse()
-        for invoice_policy in invoice_policies:
-            so_lines = (
-                self.with_context(invoice_policy=invoice_policy)
-                .filtered(lambda x, p=invoice_policy: x.order_id.invoice_policy == p)
-                .with_prefetch()
-            )
-            done_lines |= so_lines
-            if so_lines:
-                super(SaleOrderLine, so_lines)._compute_invoice_status()
-                for line in so_lines:
-                    # due to the change of context in compute methods,
-                    # assign the value in the modified context to self
-                    line_by_id[line.id].invoice_status = line.invoice_status
-        # Not to break function if (it could not happen) some records
-        # were not in so_lines
-        super(SaleOrderLine, self - done_lines)._compute_invoice_status()
+    def _compute_qty_to_invoice(self):
+        other_lines = self.env["sale.order.line"]
+        for line in self:
+            if line.product_id.type == "service" or not line.order_id.invoice_policy:
+                other_lines |= line
```

Comment:
Instead of going one by one and adding it to the recordset, which means to recreate it each time, do `filtered` for getting the result in one shot.

Reply:
done